tests/importer/importer.cpp

```diff
@@ -0,0 +1,427 @@
+    auto modelC = parser->parseModel(fileContents("resolveimports/generic.cellml"));
+
+    // ModelD will be imorted into the library as it's included as a dependency in the models here.
+    // Passing in one of the models for resolution will load modelD into the library.
+    auto importer = libcellml::Importer::create();
+    importer->resolveImports(modelA, resourcePath("resolveimports/"));
+
+    EXPECT_EQ(size_t(1), importer->libraryCount());
+
+    // Now resolve the other models and expect the library contents to be unchanged.
+    importer->resolveImports(modelB, resourcePath("resolveimports/"));
+    importer->resolveImports(modelC, resourcePath("resolveimports/"));
+
+    EXPECT_EQ(size_t(1), importer->libraryCount());
+}
+
+TEST(Importer, addModelToLibrary)
+{
+    // This test shows how a model instance can be manually added to the import library by the user,
+    // and will be used to resolve imports, rather than from an external file/url.
+    auto parser = libcellml::Parser::create();
+    auto model = parser->parseModel(fileContents("resolveimports/generic_no_source.cellml"));
+    auto importer = libcellml::Importer::create();
+
+    auto sourceModel = libcellml::Model::create("source");
+    sourceModel->addComponent(libcellml::Component::create("a"));
+    sourceModel->addComponent(libcellml::Component::create("b"));
+    sourceModel->addComponent(libcellml::Component::create("c"));
+    sourceModel->addComponent(libcellml::Component::create("d"));
+
+    // Add a model manually to the library, including the URL that it will replace in future imports.
+    EXPECT_TRUE(importer->addModel(sourceModel, "i_dont_exist.cellml"));
+
+    // Can't add to the same url key more than once.
+    EXPECT_FALSE(importer->addModel(sourceModel, "i_dont_exist.cellml"));
+
+    importer->resolveImports(model, "");
+
+    EXPECT_EQ(size_t(0), importer->issueCount());
+    EXPECT_FALSE(model->hasUnresolvedImports());
+}
+
+TEST(Importer, replaceModel)
+{
+    // This test shows how a model instance can be manually added to the import library by the user,
+    // and will be used to resolve imports, rather than from an external file/url.
+    auto parser = libcellml::Parser::create();
+    auto model = parser->parseModel(fileContents("resolveimports/generic_no_source.cellml"));
+    auto importer = libcellml::Importer::create();
+
+    auto wrongSourceModel = libcellml::Model::create("wrong");
+
+    auto rightSourceModel = libcellml::Model::create("source");
+    rightSourceModel->addComponent(libcellml::Component::create("a"));
+    rightSourceModel->addComponent(libcellml::Component::create("b"));
+    rightSourceModel->addComponent(libcellml::Component::create("c"));
+    rightSourceModel->addComponent(libcellml::Component::create("d"));
+
+    // Add a model manually to the library, including the URL that it will replace in future imports.
+    EXPECT_TRUE(importer->addModel(wrongSourceModel, "i_dont_exist.cellml"));
```

Comment:
Since I expect the model "key" to always be an absolute URL, I would encourage tests to look more like absolute URLs...e.g., "https://example.com/cellml/i_dont_exist.cellml"

Reply:
But if we're allowing the local resolution of imports (by adding models to the library manually), there's no reason that the key should be a URL, absolute or otherwise - it just needs to be a unique string? Especially since we don't require an absolute URL in the import source?  It's easy enough to change the test, but I want to understand the logic too ...

Reply:
if the key is not an absolute URL then it is not possible to resolve imports without also needing to maintain a key <==> URL mapping. The key should (MUST?) be an absolute URL.

Reply:
That's not quite how I understand it now that we have this class ... imports are resolved with a key <==> model mapping in the importer.  
In some situations, the model must be parsed from an external file/location, so it (coincidentally) needs to know an absolute URL, and absolute URLs are by their nature unique, it's just good sense to use that as the key if required.  

I would have thought that the only driving requirement for having the key as an absolute URL would be if we also required all import sources to be absolute too (so it would make printing a model easier) ... but we don't ... ?

Reply:
What we're saying is that here is a library of models. One of the purposes of such a library is to make the models available for resolving imports. To resolve imports requires that each model is associated with an absolute URL, agreed?

Import sources have an `href` which defines the source of the imported model. This is recommended to be a relative URL, in which case it needs to be resolved to an absolute URL in a specified context in order to be able to resolve the import. 

If I want to populate the library with models in order to make them available for resolving imports, I therefore need to associate each one with an absolute URL. Pretty sure. I think. At least I can't see any way they could be used otherwise.

But not sure if this is getting clearer or not...

Reply:
> What we're saying is that here is a library of models. One of the purposes of such a library is to make the models available for resolving imports. To resolve imports requires that each model is associated with an absolute URL, agreed?

Nope, this is the bit I don't agree with.  To resolve imports we need to import model(s).  The whole point of the importer is that it changes the earlier requirement away from needing absolute URLs (which was how it was before the Importer class existed) into simply requiring that the models themselves be supplied.  We don't care where they came from.  And if we don't care where they came from, we just need to be able to identify them, so we just need a unique string.  

> Import sources have an href which defines the source of the imported model. This is recommended to be a relative URL, in which case it needs to be resolved to an absolute URL in a specified context in order to be able to ~resolve the import~ **locate the model, if not in the library already**.

Maybe I'll write a complicated test and see if that makes it clearer (but don't hold your breath!)...

Reply:
> Maybe I'll write a complicated test and see if that makes it clearer (but don't hold your breath!)...

I won't hold my breath, but I think this is what will be needed to show me what I am missing here :)

Reply:
Just for completeness, I didn't hold my breath and the complicated test(s) from @kerimoyle did convince me that this is a useful way to go. The key is likely to be an absolute URL in most cases, but no requirement for it to be so and some cases where you definitely benefit from being able to make it anything..